storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java

```diff
@@ -0,0 +1,178 @@
+            GrasSimulatedUser u = users.get(0);
+            TopologyDetails td = u.getNextHighest();
+            if (td == null) {
+                users.remove(0);
+            } else {
+                double score = u.getScore(cpuAvail, memAvail, genericAvail);
+                td = u.simScheduleNextHighest();
+                LOG.info("GRAS SIM Scheduling {} with score of {}", td.getId(), score);
+                cpuAvail -= td.getTotalRequestedCpu();
+                memAvail -= (td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap());
+                for (Map.Entry<String, Double> entry : td.getTotalRequestedGenericResources().entrySet()) {
+                    String resource = entry.getKey();
+                    Double requestedAmount = entry.getValue();
+                    if (!genericAvail.containsKey(resource)) {
+                        LOG.warn("Resource: {} is not supported in this cluster. Ignoring this request.", resource);
+                    } else {
+                        genericAvail.put(resource, genericAvail.get(resource) - requestedAmount);
+                    }
+                }
+                allUserTopologies.add(td);
+            }
+        }
+        return allUserTopologies;
+    }
+
+    protected static class GrasSimulatedUser extends SimulatedUser {
+
+        // Extend support for Generic Resources in addition to CPU and Memory
+        public final Map<String, Double> guaranteedGenericResources;                // resource name : guaranteed amount
+        private Map<String, Double> assignedGenericResources = new HashMap<>();     // resource name : assigned amount
+
+        public GrasSimulatedUser(User other, ISchedulingState cluster) {
+            super(other, cluster);
+
+            Map<String, Double> guaranteedGenericResources = new HashMap<>();
+            // generic resource types that are offered
+            Set<String> availGenericResourceTypes = cluster.getClusterTotalGenericResources().keySet();
+            for (String resourceType : availGenericResourceTypes) {
+                Double guaranteedAmount = other.getGenericGuaranteed().getOrDefault(resourceType, 0.0);
+                guaranteedGenericResources.put(resourceType, guaranteedAmount);
+            }
+            this.guaranteedGenericResources = guaranteedGenericResources;
+        }
+
+        @Override
+        public TopologyDetails simScheduleNextHighest() {
+            TopologyDetails td = super.simScheduleNextHighest();
+            Map<String, Double> tdRequestedGenericResource = td.getTotalRequestedGenericResources();
+            for (Map.Entry<String, Double> entry : tdRequestedGenericResource.entrySet()) {
+                String resource = entry.getKey();
+                Double requestedAmount = entry.getValue();
+                assignedGenericResources.put(resource, assignedGenericResources.getOrDefault(resource, 0.0) + requestedAmount);
+            }
```

Comment:
should we not adjust cpu and memory as well? Very similar to how `SimulatedUser.simScheduleNextHighest()`

```                assignedCpu += td.getTotalRequestedCpu();
                assignedMemory += td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap();
```
Even better would be generally refactor this whole class to minimize the duplicate code...

Reply:
cpu and mem are adjusted by its overrided method on line 105:
```TopologyDetails td = super.simScheduleNextHighest();```

Not sure if this is what you are looking for.